pkg/login/social/azuread_oauth.go

```diff
@@ -115,7 +119,7 @@ func extractRole(claims azureClaims, autoAssignRole string) models.RoleType {
 		}
 	}
 
-	return models.ROLE_VIEWER
+	return models.RoleType("")
```

Comment:
This change feels a little fishy to me but I'm not super familiar with this area of the codebase.

If I previously had a setup where users were automatically being assigned the viewer role and I was fine with that, will that continue to work here?

I see an old test that is still passing,

```
{
			name: "Email in email claim",
			claims: &azureClaims{
				Email:             "me@example.com",
				PreferredUsername: "",
				Roles:             []string{},
				Name:              "My Name",
				ID:                "1234",
			},
			fields: fields{
				autoAssignOrgRole: "Viewer",
			},
			want: &BasicUserInfo{
				Id:      "1234",
				Name:    "My Name",
				Email:   "me@example.com",
				Login:   "me@example.com",
				Company: "",
				Role:    "Viewer",
				Groups:  []string{},
			},
		},
```

Would this have previously passed even without the `autoAssignOrgRole: "Viewer"` setting, and will that behavior change?

Reply:
No it wouldn't work. That's the main reason behind this PR, is that I was hoping to address the situation in #22741 where a user with no role assigned in AAD shouldn't be authorized to access Grafana, instead of being granted the viewer role automatically. 

To answer the second question, yes the behavior would change. It would have passed without the setting but with the proposed change in the PR, it wouldn't.

Reply:
> where a user with no role assigned in AAD shouldn't be authorized to access Grafana, instead of being granted the viewer role automatically

This should only be true when the new strict mode flag is enabled though, correct? If it is not enabled then I believe the previous behavior should stay the same. I agree with your assessment, it should not be that way, but we also should not break existing behavior.

I'll defer to @vtorosyan on this, but I think that we would be prefer a warning log if they reach this point where the user is being assigned the default role of VIEWER and strict mode is disabled

Reply:
Now that I think about it, the strict mode flag doesn't really do anything in the azure_oauth code. In the [extractRole](https://github.com/grafana/grafana/blob/main/pkg/login/social/azuread_oauth.go#L126-L136) method, we only return roles that match admin, viewer or editor. The strict mode check pretty much does the same thing. The code I based this off in `okta_oauth.go` needed the strict mode since its [`extractRole` method](https://github.com/grafana/grafana/blob/main/pkg/login/social/okta_oauth.go#L123-L133) grabs any role from the claims token.

That said, the PR change will just be changing the default role from Viewer to nothing. Let's wait on their response to determine the final direction.

Reply:
Yes, `extractRole()` function returns Viewer role by defaut, so it's always valid. I think code that checks strict option should be in that function. Also, pay attention to the `auto_assign_org_role` (`autoAssignRole` inside `extractRole()`) config that allows to return another role by default.

Reply:
I'm not really sure what to do here. If we want the `extractRole` to be always valid, I don't see a way to deny a user with no role assigned in AAD, regardless of strict mode. The [only workaround](https://github.com/grafana/grafana/issues/22741#issuecomment-598233882) for this issue doesn't work in [all cases ](https://github.com/grafana/grafana/issues/22741#issuecomment-598751235). Any suggestions on how to handle this situation, without modifying the return of default role?

Reply:
```
func extractRole(claims azureClaims, autoAssignRole string, strictMode bool) models.RoleType {
	if len(claims.Roles) == 0 {
		return models.RoleType(autoAssignRole)
	}
	roleOrder := []models.RoleType{
		models.ROLE_ADMIN,
		models.ROLE_EDITOR,
		models.ROLE_VIEWER,
	}
	for _, role := range roleOrder {
		if found := hasRole(claims.Roles, role); found {
			return role
 		}
 	}

 	if strictMode {
 	 return models.RoleType("")
 	}
 	return models.ROLE_VIEWER
}
```

(or better yet, return an `error` rather than relying on `""` being an invalid role)?

We're _very_ unlikely to merge a change to the default behavior, that would be a breaking change ("Click Manifest, then define the required Application Role values for Grafana: Viewer, Editor, or Admin. If not defined, all users will have the Viewer role." from the docs implies that we _at least_ would need to add an exception for `""->Viewer`).

Reply:
@sivamu Is this (and the below feedback) enough to get this working and merged?